### PR TITLE
tests: benchmarks: sys_kernel: Add mem_slab benchmark and changed checking for pending threads algo

### PR DESCRIPTION
**Pull request by kamlaz**
In single-thread applications or when mem_slab is used only by one thread,
there is no need to check if any therad is pending during freeing the slab.
In multithreading application this can be optimized - check for
pending threads (in k_mem_slab_free) could be done only if there is no memory
to allocate.
This change introduces significant time savings.
Execution mem_slab benchmark on nrf5340 (App core @64MHz), 1000 iterations.
Times per single alloc/free:
Without optimization:
Alloc: 1068 [ns]
Free: 1617 [ns]

With optimization
Alloc: 1068 [ns]
Free: 915 [ns]

Difference is significant: 702[ns].

### CODE REVIEW

File: tests/benchmarks/sys_kernel/src/mem_slab.c [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+/*
+ * Copyright (c) 2020, Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "syskernel.h"
+
+#define MEM_SLAB_BLOCK_SIZE  (32)
+#define MEM_SLAB_BLOCK_CNT   (10240)
```
Comment (nordic-krch) [inline tests/benchmarks/sys_kernel/src/mem_slab.c:12]: this is probably too aggressive as most boards won't have enough ram. `BLOCK_SIZE` could be 8. You can play with block count to get fairly consistent numbers. Maybe 500 would be enough.

File: tests/benchmarks/sys_kernel/src/mem_slab.c [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+		"Memslab #1");
+	fprintf(output_file, sz_description,
+		"\n\tk_mem_slab_alloc");
+	printf(sz_test_start_fmt);
+
+	t = BENCH_START();
+	i = mem_slab_alloc_test(number_of_loops, test_repeats);
+	t = TIME_STAMP_DELTA_GET(t);
+
+	return_value += check_result(i, t);
```
Comment (nordic-krch) [inline tests/benchmarks/sys_kernel/src/mem_slab.c:95]: i think that ideally at the end it should print an avarage time for single alloc and free call.

File: tests/benchmarks/sys_kernel/src/mem_slab.c (hunk 1 above)
Comment (nordic-krch) [reply tests/benchmarks/sys_kernel/src/mem_slab.c:12]: i tried 8/256 and modified a bit test to give avg time per single call and it seems that results are fairly comparable with 512 or 1024 iterations.

**Comment by nordic-krch** [comment]
btw for `qemu_cortex_m3` i got following results for single `k_mem_slab_free`:
- without optimization: 4268ns
- with: 2468ns

File: tests/benchmarks/sys_kernel/src/mem_slab.c (hunk 1 above)
Comment (kamlaz) [reply tests/benchmarks/sys_kernel/src/mem_slab.c]: I modified a test a little. Now it runs `number_of_loops`, which is more obvious. Also reduced the size of single chunk in slab to `8`.

File: tests/benchmarks/sys_kernel/src/mem_slab.c (hunk 2 above)
Comment (kamlaz) [reply tests/benchmarks/sys_kernel/src/mem_slab.c:86]: Now it outputs allocation and freeing time per single operation.

**Review by nordic-krch — APPROVED** [review]
Looks good. Can you update PR description with results for single alloc and free?

**Comment by nordic-krch** [comment]
@andyross @andrewboie can you take a look?

File: include/kernel.h [hunk 3]
```diff
@@ -4224,6 +4224,7 @@ struct k_mem_slab {
 	char *free_list;
 	uint32_t num_used;
 
+	bool is_out_of_mem;
```
Comment (anangl) [inline include/kernel.h:4227]: It this new field actually needed? Couldn't `free_list == NULL` be checked instead?

File: tests/benchmarks/sys_kernel/src/mem_slab.c [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+ * @param test_repeats Amount of test repeats per loop.
+ *
+ * @return NUmber of done loops.
+ */
+static int mem_slab_alloc_test(int no_of_loops)
+{
+	int i;
+
+	for (i = 0; i < no_of_loops; i++) {
+		if (k_mem_slab_alloc(&my_slab, &slab_array[i], K_MSEC(100))
```
Comment (anangl) [inline tests/benchmarks/sys_kernel/src/mem_slab.c:38]: Wouldn't `K_NO_WAIT` make more sense in the benchmarking context?

**Review by anangl — APPROVED** [review]

**Comment by kamlaz** [comment]
@andyross @andrewboie would you mind taking a look please?

File: tests/benchmarks/sys_kernel/src/mem_slab.c [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "syskernel.h"
+
+#define MEM_SLAB_BLOCK_SIZE  (8)
+#define MEM_SLAB_BLOCK_CNT   (NUMBER_OF_LOOPS)
+#define MEM_SLAB_BLOCK_ALIGN (4)
+
+K_MEM_SLAB_DEFINE(my_slab,
```
[the remark this answers is not in the document]
Comment (kamlaz) [reply tests/benchmarks/sys_kernel/src/mem_slab.c]: Added.

File: tests/benchmarks/sys_kernel/src/mem_slab.c [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+		  MEM_SLAB_BLOCK_SIZE,
+		  MEM_SLAB_BLOCK_CNT,
+		  MEM_SLAB_BLOCK_ALIGN);
+
+/* Array contains pointers to allocated regions. */
+void *slab_array[MEM_SLAB_BLOCK_CNT];
```
[the remark this answers is not in the document]
Comment (kamlaz) [reply tests/benchmarks/sys_kernel/src/mem_slab.c]: Added.

File: include/kernel.h (hunk 3 above)
Comment (kamlaz) [reply include/kernel.h]: Changed.

File: tests/benchmarks/sys_kernel/src/mem_slab.c (hunk 4 above)
Comment (kamlaz) [reply tests/benchmarks/sys_kernel/src/mem_slab.c]: Changed.

**Comment by kamlaz** [comment]
@carlescufi  Can we merge this?

**Pull request merged**